Add starknet doc (#32)
Cypher-Laboratory/Alice-s-Ring: packages/documentation/docs/SAG/SAG-for-starknet.md

---
id: SAG-starknet
title: "SAG for Starknet"
---

# SAG for Starknet

> The **SAG for Starknet** implementation enables verification of Spontaneous Anonymous Group (SAG) signatures on the Starknet platform. This implementation leverages Starknet-specific optimizations to perform ring signature verification efficiently on-chain, supporting both SECP256K1 and ED25519 elliptic curves.

Reviewer: ```suggestion
The **SAG for Starknet** implementation enables verification of Spontaneous Anonymous Group (SAG) signatures on the Starknet blockchain. This implementation leverages Starknet-specific optimizations to perform ring signature verification efficiently on-chain, supporting both SECP256K1 and ED25519 elliptic curves.
```

The SAG verification functionality on Starknet is provided by the `starknet-sag-ts` package from [Alice's Ring](https://github.com/Cypher-Laboratory/Alice-s-Ring-SAG-TS), a TypeScript library for creating SAG signatures.

## Key Differences in the Starknet Implementation

The Starknet version of the SAG verifier shares the same fundamental logic as the EVM version, with a few important adjustments to optimize for the Starknet ecosystem:

1. **Poseidon Hash Function**: The signer and verifier use the Poseidon hashing function, which is computationally cheaper on Starknet compared to traditional hashing algorithms.

2. **Garaga for Multi-Scalar Multiplication (MSM)**: The implementation is using Garaga allowing efficient on-chain multi-scalar multiplication (MSM). The process incorporates hints passed alongside the ring signature to perform MSM efficiently.

## Generating a Ring Signature for Starknet

Here's how to generate a ring signature for starknet using Alice's Ring in TypeScript:

```typescript
import {
  RingSignature,
  Curve,
  CurveName,
  Point,
} from "@cypher-laboratory/alicesring-sag-starknet";

const curve = new Curve(CurveName.SECP256K1);

> const ring: Point[] = []; // Your ring of public keys

Reviewer: might be good to add 2 or 3 public keys so the user can see how they should be encoded (compressed to 33Bytes I guess)

Author: done

const message = "Hello, Alice's Ring!";

const signerPrivateKey = BigInt("0x..."); // The signer's private key
async function ringSignForStarknet() {
  // Sign
  const signature = RingSignature.sign(ring, signerPrivateKey, message, curve);

  // Verify locally
  console.log("Is signature valid?", signature.verify()); // Should output: true

  const signature = RingSignature.sign(ring, signerPrivKey, message, curve);
  const rawCallData = await signature.getCallData();

>   console.log(rawCallData);

Reviewer: can you add some context in the log ? So when someone run the code, they can easily know what this is.

Author: done

> async function ringSignForStarknet() {
>   // Sign
>   const signature = RingSignature.sign(ring, signerPrivateKey, message, curve);
> 
>   // Verify locally
>   console.log("Is signature valid?", signature.verify()); // Should output: true
> 
>   const signature = RingSignature.sign(ring, signerPrivKey, message, curve);
>   const rawCallData = await signature.getCallData();
>   console.log(rawCallData);
> }
> ringSignForStarknet(); 

Reviewer: ```suggestion

// Sign
const signature = RingSignature.sign(ring, signerPrivateKey, message, curve);

// Verify locally
console.log("Is signature valid?", signature.verify()); // Should output: true

(async () => {
  // Get the signature data to send to the contract
  const signature = RingSignature.sign(ring, signerPrivateKey, message, curve);
  const rawCallData = await signature.getCallData();
  console.log(rawCallData);
})();
```

- snippet was not valid (signature was defined twice in the same block and `signerPrivKey` was undefined)
- replaced the function by a simple async block (increase readability)
- only let the async call in the async block
- explained what is done in the `rawCallData` generation 

Can you also explain a little more the link between `rawCallData` and the contract (how is it used, if it is a function argument, explain whch one, ...). Maybe add a snippet where you generate a signature and verify it using the contract

```

> **Important:** ringSignature.getCallData is an **async** function.

Reviewer: ```suggestion
**Important:** ringSignature.getCallData() is an **async** function.
```

## Verifying a Ring Signature On-Chain

> To verify the ring signature on-chain, you can use the [SAG verifier in Cairo](https://github.com/Cypher-Laboratory/Alice-s-Ring-Cairo-verifier).

Reviewer: why isn't this verifier in the monorepo ?

Author: I don't know how scarb will manage the import of the package in a monorepo (as you pass the git repo to import). 
But we could try !

Below is an example of how to integrate the verifier into your Cairo contract.

### Installation

Add the SAG Cairo Verifier to your **Scarb.toml**:

```toml
[dependencies]
alices_ring_cairo_verifier = {git = "https://github.com/Cypher-Laboratory/Alice-s-Ring-Cairo-verifier.git"}
```

### Importing the Verifier Interface

```cairo
use alices_ring_cairo_verifier::structtype::ringsignature;

#[starknet::interface]
pub trait IRingSignatureVerifier<TContractState> {
    fn verify_rs(self: @TContractState, ring_signature: RingSignature) -> bool;
}

#[starknet::contract]
mod RingSignatureVerifier {
    use alices_ring_cairo_verifier::verify;
    use alices_ring_cairo_verifier::structType::RingSignature;

    #[storage]
    struct Storage {}

    #[abi(embed_v0)]
    impl IRingSignatureVerifier of super::IRingSignatureVerifier<ContractState> {
        fn verify_rs(self: @ContractState, ring_signature: RingSignature) -> bool {
            verify(ring_signature)
        }
    }
}
```

> ### Using Voyager block explorer :
> 
> You can also pass the raw call data to the deployed contract at the following address:
> 
> **Starknet Contract Address**: [`0x0317Ce745DE0A65308A19f6e184Be43A17F20D1eAe04B7433d984Df1D571DAA4`](https://sepolia.voyager.online/contract/0x0317Ce745DE0A65308A19f6e184Be43A17F20D1eAe04B7433d984Df1D571DAA4#readContract) on Sepolia Starknet.
> 

Reviewer: ```suggestion
```

this is defined twice (see `Deployed address` section)

Author: this one refers to the action of making a call data to the block explorer (similar to this : https://garaga.gitbook.io/garaga/using-garaga-libraries-in-your-cairo-project/ec-multi-scalar-multiplication)

Reviewer: can you explain this in the doc ?

Author: yep i will do a small video like in the garaga book

### Number of steps

> Verifying a SAG signature requires approximately 46,800 steps per member in the ring.

Reviewer: ```suggestion
Verifying a SAG signature requires approximately 46,800 steps per public key in the ring.
```

## Deployed address

| Network          |                                                                                                        | Deployed Address                                                                                                                                                                   |
| ---------------- | ------------------------------------------------------------------------------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Starknet-Sepolia | <img src="https://cryptologos.cc/logos/starknet-token-strk-logo.png?v=035" alt="Starknet" width="30"/> | [`0x0317Ce745DE0A65308A19f6e184Be43A17F20D1eAe04B7433d984Df1D571DAA4`](https://sepolia.voyager.online/contract/0x0317Ce745DE0A65308A19f6e184Be43A17F20D1eAe04B7433d984Df1D571DAA4) |

## Contribution

We welcome contributions! To contribute, follow these steps:

1. Fork the [Repository for the signer](https://github.com/Cypher-Laboratory/Alice-s-Ring)

>    or the [Repository for the verifier](https://github.com/Cypher-Laboratory/Alice-s-Ring-Cairo-verifier) : Create a fork of the repository in your GitHub account.

Reviewer: Why aren't them both in the `Alice's Ring` monorepo ?

## Key Differences in the Starknet Implementation

The Starknet version of the SAG verifier shares the same fundamental logic as the EVM version, with a few important adjustments to optimize for the Starknet ecosystem:

1. **Poseidon Hash Function**: The signer and verifier use the Poseidon hashing function, which is computationally cheaper on Starknet compared to traditional hashing algorithms.

2. **Garaga for Multi-Scalar Multiplication (MSM)**: The implementation is using Garaga allowing efficient on-chain multi-scalar multiplication (MSM). The process incorporates hints passed alongside the ring signature to perform MSM efficiently.

## Generating a Ring Signature for Starknet

Here's how to generate a ring signature for starknet using Alice's Ring in TypeScript:

```typescript
import {
  RingSignature,
  Curve,
  CurveName,
  Point,
} from "@cypher-laboratory/alicesring-sag-starknet";

const curve = new Curve(CurveName.SECP256K1);
const message = "Hello, Alice's Ring!";

// Define an array of compressed public keys

> const compressedPublickeys = [
>   "0316d7da70ba247a6a40bb310187e8789b80c45fa6dc0061abb8ced49cbe7f887f",
>   "0221869ca3ae33be3a7327e9a0272203afa72c52a5460ceb9f4a50930531bd926a",
>   "02337d6f577e66a21a7831c087c6836a1bae37086bf431400811ac7c6e96c8ccbb",
> ];

Reviewer: ```suggestion
const compressedPublickeys = [
  "0316d7da70ba247a6a40bb310187e8789b80c45fa6dc0061abb8ced49cbe7f887f",
  "0221869ca3ae33be3a7327e9a0272203afa72c52a5460ceb9f4a50930531bd926a",
  // ... 
];
```
